cmds/create-ffs/create-ffs.go

```diff
@@ -0,0 +1,203 @@
+
+	fType  uefi.FVFileType
+	fGUID  *guid.GUID
+	depOps []uefi.DepExOp
+	printf = func(string, ...interface{}) {}
+)
+
+const (
+	usageString = "Usage: create-ffs [flags] file.efi [...]"
+)
+
+func createDepExes(deps string) ([]uefi.DepExOp, error) {
+	var err error
+	ops := []uefi.DepExOp{}
+
+	if strings.ToUpper(deps) == "TRUE" {
+		// Create just "TRUE" and "END" for now, but this feels unnecessary to me.
+		ops = append(ops, uefi.DepExOp{OpCode: "TRUE"})
+		ops = append(ops, uefi.DepExOp{OpCode: "END"})
+		return ops, nil
+	}
+
+	// we expect a space-separated list of GUIDs
+	guids := strings.Split(deps, " ")
+	numGUIDS := len(guids)
+	for _, guidStr := range guids {
+		var g *guid.GUID
+
+		if g, err = guid.Parse(guidStr); err != nil {
+			return nil, err
+		}
+		printf("depex guid requested: %v", *g)
+		ops = append(ops, uefi.DepExOp{OpCode: "PUSH", GUID: g})
+	}
+
+	// Append an "AND" for n-1 pushes.
+	for i := 1; i < numGUIDS; i++ {
+		ops = append(ops, uefi.DepExOp{OpCode: "AND"})
+	}
+
+	ops = append(ops, uefi.DepExOp{OpCode: "END"})
+	return ops, nil
+}
+
+func parseFlags() error {
+	var ok bool
+	var err error
+
+	if *debug {
+		printf = log.Printf
+	}
+
+	// Check filetypes
+	fType, ok = uefi.NamesToFileType[*filetype]
+	if !ok {
+		validTypes := []string{}
+		for k := range uefi.NamesToFileType {
+			validTypes = append(validTypes, k)
+		}
+		return fmt.Errorf("unable to get EFI File type, got %v, expected values in\n{%v}",
+			*filetype, strings.Join(validTypes, ", "))
+	}
+
+	if *guidString != "" {
+		fGUID, err = guid.Parse(*guidString)
+		if err != nil {
+			return err
+		}
+	} else if *name != "" {
+		// We sha1 the name to get a reproducible GUID.
+		fGUID = &guid.GUID{}
+		sum := sha1.Sum([]byte(*name))
+		copy(fGUID[:], sum[:guid.Size])
+	} else {
+		return errors.New("no GUID or name provided, please provide at least one")
+	}
+
+	if *outfile == "" {
+		return errors.New("we don't currently support dumping to stdout, please specify an output file")
+	}
+
+	if *depex != "" {
+		depOps, err = createDepExes(*depex)
+		if err != nil {
+			return fmt.Errorf("can't parse depex guids, got %v", err)
+		}
+	}
+
+	return nil
+}
+
+func usage() {
+	log.Print(usageString)
+	flag.PrintDefaults()
+	os.Exit(1)
+}
+
+func main() {
+	flag.Parse()
+	if err := parseFlags(); err != nil {
+		log.Fatal(err)
+	}
+
+	args := flag.Args()
+	if alen := len(args); alen < 1 {
+		usage()
+	}
+
+	secData, err := ioutil.ReadFile(args[0])
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	printf("type requested: %v", fType)
+	printf("name requested: %v", *name)
+	printf("version requested: %v", *version)
+	printf("guid: %v", fGUID)
+
+	secType := uefi.SectionTypeRaw
+	switch fType {
+	case uefi.FVFileTypeApplication, uefi.FVFileTypeCombinedSMMDXE,
+		uefi.FVFileTypeCombinedPEIMDriver, uefi.FVFileTypeDriver:
+		secType = uefi.SectionTypePE32
+	}
+
+	file := &uefi.File{}
+	file.Header.Type = fType
+	file.Header.GUID = *fGUID
+
```

Comment:
Here I missed the following lines to build an image equivalent to what the perl version is doing:
```
	file.Header.Attributes = 0x40
	file.Header.State = 0xF8
```
Presumably the second line is due to #263 since I tested after rebasing.

With the above change it became a drop-in replacement to perl create-ffs !

Reply:
The attribute is not necessary, it might be better to have it as parameter

Reply:
I will add in the attributes as well, we can make it a parameter later.